File: ios/Classes/DixaMessengerFlutterPlugin.swift
import Flutter
import UIKit
import DixaMessenger

public class DixaMessengerFlutterPlugin: NSObject, FlutterPlugin {
    private var instanceChannels: [String: FlutterMethodChannel] = [:]
    private var instanceConfigs: [String: DixaConfiguration] = [:]
    private var registrar: FlutterPluginRegistrar?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "dixa_messenger_flutter", binaryMessenger: registrar.messenger())
        let instance = DixaMessengerFlutterPlugin()
        instance.registrar = registrar
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let methodNotImplemented = FlutterMethodNotImplemented
        
        switch call.method {
        case "createInstance":
            Task { @MainActor in
                await createInstance(call, result: result)
            }
        case "removeInstance":
            Task { @MainActor in
                await removeInstance(call, result: result)
            }
        default:
            result(methodNotImplemented)
        }
    }
    
    @MainActor
    private func createInstance(_ call: FlutterMethodCall, result: @escaping FlutterResult) async {
        guard let args = call.arguments as? [String: Any],
              let instanceName = args["instanceName"] as? String,
              let configMap = args["config"] as? [String: Any],
              let apiKey = configMap["apiKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }
        
        // Create configuration
        var config = DixaConfiguration().apikey(apiKey)
        
        // Set log level
        if let logLevelStr = configMap["logLevel"] as? String {
            let logLevel: LogLevel
            switch logLevelStr {
            case "error": logLevel = .error
            case "warning": logLevel = .warning
            case "all": logLevel = .all
            default: logLevel = .none
            }
            config = config.logLevel(logLevel)
        }
        
        // Set push environment
        if let pushEnvStr = configMap["pushEnvironment"] as? String {
            let pushEnv: DixaConfiguration.PushEnvironment = pushEnvStr == "sandbox" ? .sandbox : .production
            config = config.pushEnvironment(pushEnv)
        }
        
        // Set supported languages
        if let supportedLanguages = configMap["supportedLanguages"] as? [String] {
            config = config.supportedLanguages(supportedLanguages)
        }
        
        // Store configuration for this instance
        instanceConfigs[instanceName] = config
        
        // Configure messenger for this instance
        await Messenger.configure(config)
        
        // Handle authentication
        if let authConfig = configMap["authentication"] as? [String: Any],
           let authType = authConfig["type"] as? String {
            switch authType {
            case "claimed":
                if let username = authConfig["username"] as? String,
                   let email = authConfig["email"] as? String {
                    await Messenger.updateUserCredentials(username: username, email: email)
                }
            case "verified":
                if let token = authConfig["verificationToken"] as? String {
                    await Messenger.verifyUser(with: token)
                }
            default:
                break
            }
        }
        
        // Create instance-specific channel
        guard let registrar = registrar else {
            result(FlutterError(code: "NO_REGISTRAR", message: "No registrar available", details: nil))
            return
        }
        
        let instanceChannel = FlutterMethodChannel(
            name: "dixa_messenger_flutter/\(instanceName)",
            binaryMessenger: registrar.messenger()
        )
        
        instanceChannel.setMethodCallHandler { [weak self] call, result in
            Task {
                await self?.handleInstanceMethod(instanceName: instanceName, call: call, result: result)
            }
        }
        
        instanceChannels[instanceName] = instanceChannel
        result(nil)
    }
    
    @MainActor
    private func removeInstance(_ call: FlutterMethodCall, result: @escaping FlutterResult) async {
        guard let args = call.arguments as? [String: Any],
              let instanceName = args["instanceName"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }
        
        instanceChannels[instanceName]?.setMethodCallHandler(nil)
        instanceChannels.removeValue(forKey: instanceName)
        instanceConfigs.removeValue(forKey: instanceName)
        result(nil)
    }
    
    @MainActor
    private func handleInstanceMethod(instanceName: String, call: FlutterMethodCall, result: @escaping FlutterResult) async {
        let methodNotImplemented = FlutterMethodNotImplemented
        
        // Ensure we have the correct configuration for this instance
        guard let config = instanceConfigs[instanceName] else {
            result(FlutterError(code: "NO_CONFIG", message: "No configuration found for instance", details: nil))
            return
        }
        
        // Reconfigure messenger with the correct configuration
        await Messenger.configure(config)
        
        switch call.method {
        case "openMessenger":
            guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                  let rootViewController = windowScene.windows.first?.rootViewController else {
                result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "No root view controller", details: nil))
                return
            }
            await Messenger.openMessenger(from: rootViewController)
            result(nil)
            
        case "updateUserCredentials":
            guard let args = call.arguments as? [String: Any],
                  let username = args["username"] as? String,
                  let email = args["email"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            await Messenger.updateUserCredentials(username: username, email: email)
            result(nil)
            
        case "setVerificationToken":
            guard let args = call.arguments as? [String: Any],
                  let token = args["token"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
                return
            }
            await Messenger.verifyUser(with: token)
            result(nil)
            
        case "clearUserCredentials":
            await Messenger.clearUserCredentials()
            result(nil)
            
        case "clearVerificationToken":
            await Messenger.clearVerificationToken()
            result(nil)
            
        case "setUnreadMessagesCountListener":
            await Messenger.unreadMessagesCountListener { [weak self] count in
                self?.instanceChannels[instanceName]?.invokeMethod("onUnreadCountChanged", arguments: ["count": count])
            }
            result(nil)
            
        case "registerPushNotificationToken":
            guard let args = call.arguments as? [String: Any],
                  let tokenStr = args["token"] as? String,
                  let tokenData = tokenStr.data(using: .utf8) else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid token", details: nil))
                return
            }
            await Messenger.pushNotification.register(deviceToken: tokenData)
            result(nil)
            
        default:
            result(methodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/example/dixa_messenger_flutter/DixaMessengerFlutterPlugin.kt
package com.example.dixa_messenger_flutter

import android.app.Activity
import android.app.Application
import android.content.Context
import com.dixa.messenger.DixaMessenger
import com.dixa.messenger.LogLevel
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ConcurrentHashMap

class DixaMessengerFlutterPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var binaryMessenger: io.flutter.plugin.common.BinaryMessenger
    private var activity: Activity? = null
    private val instanceChannels = ConcurrentHashMap<String, MethodChannel>()
    private val instanceConfigs = ConcurrentHashMap<String, DixaMessenger.Configuration>()
    private var currentInstance: String? = null
    
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        binaryMessenger = flutterPluginBinding.binaryMessenger
        channel = MethodChannel(binaryMessenger, "dixa_messenger_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "createInstance" -> createInstance(call, result)
            "removeInstance" -> removeInstance(call, result)
            else -> result.notImplemented()
        }
    }
    
    private fun createInstance(call: MethodCall, result: Result) {
        val instanceName = call.argument<String>("instanceName")!!
        val configMap = call.argument<Map<String, Any>>("config")!!
        
        try {
            // Create configuration
            val configBuilder = DixaMessenger.Configuration.Builder()
                .setApiKey(configMap["apiKey"] as String)
            
            // Set log level
            val logLevelStr = configMap["logLevel"] as? String
            logLevelStr?.let {
                val logLevel = when(it) {
                    "error" -> LogLevel.ERROR
                    "warning" -> LogLevel.WARNING
                    "all" -> LogLevel.ALL
                    else -> LogLevel.NONE
                }
                configBuilder.setLogLevel(logLevel)
            }
            
            // Set preferred language
            val preferredLanguage = configMap["preferredLanguage"] as? String
            preferredLanguage?.let {
                configBuilder.setPreferredLanguage(it)
            }
            
            // Set authentication
            val authConfig = configMap["authentication"] as? Map<String, Any>
            authConfig?.let { auth ->
                when (auth["type"] as String) {
                    "claimed" -> {
                        val username = auth["username"] as String
                        val email = auth["email"] as String
                        configBuilder.setUserCredentials(username, email)
                    }
                    "verified" -> {
                        val token = auth["verificationToken"] as String
                        configBuilder.setVerificationToken(token)
                    }
                }
            }
            
            val config = configBuilder.build()
            instanceConfigs[instanceName] = config
            
            // Initialize DixaMessenger for this instance
            activity?.let { act ->
                if (currentInstance == null) {
                    DixaMessenger.init(config, act.application)
                    currentInstance = instanceName
                }
                
                // Create instance-specific channel
                val instanceChannel = MethodChannel(
                    binaryMessenger,
                    "dixa_messenger_flutter/$instanceName"
                )
                instanceChannel.setMethodCallHandler { call, result ->
                    handleInstanceMethod(instanceName, call, result)
                }
                instanceChannels[instanceName] = instanceChannel
                
                result.success(null)
            } ?: result.error("NO_ACTIVITY", "Activity not available", null)
            
        } catch (e: Exception) {
            result.error("INIT_ERROR", "Failed to initialize Dixa Messenger", e.message)
        }
    }
    
    private fun removeInstance(call: MethodCall, result: Result) {
        val instanceName = call.argument<String>("instanceName")!!
        instanceChannels.remove(instanceName)?.setMethodCallHandler(null)
        result.success(null)
    }
    
    private fun switchToInstance(instanceName: String) {
        val config = instanceConfigs[instanceName] ?: return
        activity?.let { act ->
            DixaMessenger.init(config, act.application)
            currentInstance = instanceName
        }
    }
    
    private fun handleInstanceMethod(instanceName: String, call: MethodCall, result: Result) {
        // Switch to the correct instance before performing any operation
        if (currentInstance != instanceName) {
            switchToInstance(instanceName)
        }

        when (call.method) {
            "openMessenger" -> {
                activity?.let { act ->
                    DixaMessenger.openMessenger(act)
                    result.success(null)
                } ?: result.error("NO_ACTIVITY", "Activity not available", null)
            }
            "openMessengerSheet" -> {
                val offsetPx = call.argument<Int>("offsetPx") ?: 0
                activity?.let { act ->
                    DixaMessenger.openMessenger(act)
                    result.success(null)
                } ?: result.error("NO_ACTIVITY", "Activity not available", null)
            }
            "updateUserCredentials" -> {
                val username = call.argument<String>("username")!!
                val email = call.argument<String>("email")!!
                DixaMessenger.updateUserCredentials(username, email)
                result.success(null)
            }
            "setVerificationToken" -> {
                val token = call.argument<String>("token")!!
                // Note: Direct token update not supported, need to reinitialize
                result.error("NOT_SUPPORTED", "Direct token update not supported", null)
            }
            "clearUserCredentials" -> {
                DixaMessenger.clearUserCredentials()
                result.success(null)
            }
            "clearVerificationToken" -> {
                DixaMessenger.clearVerificationToken()
                result.success(null)
            }
            "setUnreadMessagesCountListener" -> {
                DixaMessenger.setUnreadMessagesCountListener { count ->
                    instanceChannels[instanceName]?.invokeMethod("onUnreadCountChanged", 
                        mapOf("count" to count))
                }
                result.success(null)
            }
            "registerPushNotificationToken" -> {
                val token = call.argument<String>("token")!!
                DixaMessenger.PushNotifications.registerNewToken(token)
                result.success(null)
            }
            "processPushNotification" -> {
                // This would need RemoteMessage object, simplified for now
                result.success(false)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        instanceChannels.values.forEach { it.setMethodCallHandler(null) }
        instanceChannels.clear()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }
}